File: src/conventions/pipeline/binding-resolver.ts
/**
 * Canonical binding resolution for $-prefixed placeholders in fact IDs.
 *
 * Surface clauses use parameterized fact IDs like `hand.suitLength.$suit`
 * with surfaceBindings `{ suit: "hearts" }`. This module provides the
 * single source of truth for resolving those placeholders into concrete
 * fact IDs before any downstream lookup.
 */

import type { BidMeaningClause } from "../../core/contracts/meaning";

/**
 * Resolve $-prefixed binding references in a factId.
 * Unresolved keys are left as-is (fail-closed).
 */
export function resolveFactId(
  factId: string,
  bindings?: Readonly<Record<string, string>>,
): string {
  if (!bindings) return factId;
  return factId.replace(/\$(\w+)/g, (_, key: string) => {
    const value = bindings[key];
    return value !== undefined ? value : `$${key}`;
  });
}

/**
 * Resolve bindings in a BidMeaningClause, returning a new clause
 * with the resolved factId. Returns the same object if no resolution needed.
 */
export function resolveClause(
  clause: BidMeaningClause,
  bindings?: Readonly<Record<string, string>>,
): BidMeaningClause {
  if (!bindings) return clause;
  const resolved = resolveFactId(clause.factId, bindings);
  return resolved === clause.factId ? clause : { ...clause, factId: resolved };
}

File: src-tauri/crates/bridge-conventions/src/pipeline/evaluation/binding_resolver.rs
//! Binding resolver — resolves $-prefixed placeholders in fact IDs.
//!
//! Mirrors TS from `pipeline/evaluation/binding-resolver.ts`.

use std::collections::HashMap;

use crate::types::meaning::BidMeaningClause;

/// Resolve a fact ID by replacing $-prefixed placeholders with binding values.
pub fn resolve_fact_id(fact_id: &str, bindings: &HashMap<String, String>) -> String {
    let mut result = fact_id.to_string();
    for (key, value) in bindings {
        let placeholder = format!("${}", key);
        result = result.replace(&placeholder, value);
    }
    result
}

/// Resolve bindings in a BidMeaningClause, returning a new clause with resolved fact_id.
pub fn resolve_clause(
    clause: &BidMeaningClause,
    bindings: &HashMap<String, String>,
) -> BidMeaningClause {
    BidMeaningClause {
        fact_id: resolve_fact_id(&clause.fact_id, bindings),
        operator: clause.operator,
        value: clause.value.clone(),
        clause_id: clause.clause_id.clone(),
        description: clause.description.clone(),
        rationale: clause.rationale.clone(),
        is_public: clause.is_public,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn resolve_no_placeholders() {
        let bindings = HashMap::new();
        assert_eq!(resolve_fact_id("hand.hcp", &bindings), "hand.hcp");
    }

    #[test]
    fn resolve_single_placeholder() {
        let mut bindings = HashMap::new();
        bindings.insert("suit".into(), "hearts".into());
        assert_eq!(
            resolve_fact_id("hand.suitLength.$suit", &bindings),
            "hand.suitLength.hearts"
        );
    }
}
